crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -1725,9 +1725,73 @@ impl PgManager {
                 },
             ));
         }
-
         Ok(Some(connection))
     }
+
+    pub(crate) async fn fetch_dynamic_field_object(
+        &self,
+        address: SuiAddress,
+        name: DynamicFieldName,
+    ) -> Result<Option<DynamicField>, Error> {
+        let type_tag =
+            TypeTag::from_str(&name.type_).map_err(|e| Error::Internal(format!("{e}")))?;
+        // This is expected to be from DynamicField.name.bcs, and is the undecorated MoveValue K of DF<K, V>
+        // Or for dynamic objects, the inner type of DF<Wrapper<T>, V>
+        let name_bcs_value = &name.bcs.0;
+        let parent_object_id =
+            ObjectID::from_bytes(address.as_slice()).map_err(|e| Error::Internal(e.to_string()))?;
+        let id = sui_types::dynamic_field::derive_dynamic_field_id(
+            parent_object_id,
+            &type_tag,
+            name_bcs_value,
+        )
+        .expect("oops");
+
+        println!("id: {:?}", id);
+
+        let stored_obj = self.get_obj(id.to_vec(), None).await?;
+        if let Some(stored_object) = stored_obj {
+            let df_object_id = stored_object.df_object_id.as_ref().ok_or_else(|| {
+                Error::Internal("Dynamic field does not have df_object_id".to_string())
+            })?;
+            let df_object_id = SuiAddress::from_bytes(df_object_id)
+                .map_err(|e| Error::Internal(format!("{e}")))?;
+            println!("df_object_id: {:?}", df_object_id);
+            return Ok(Some(DynamicField {
+                stored_object,
+                df_object_id,
+                df_kind: DynamicFieldType::DynamicField,
+            }));
+        }
+
+        // Try reading as dynamic field object by wrapping the inner T back into Wrapper<T> of DF<Wrapper<T>, V>
```

Comment:
Hmm, this is not great -- that we try the non-object case, and if that doesn't work, we try the object case.  It means that results that appear in `dynamicFieldConnection` may be inaccessible via a point-lookup API, and it also means that accessing dynamic object fields involve more lookups.

I would prefer that we revisit the schema and expose separate APIs -- for DynamicField and DynamicObjectField.

Reply:
I confused myself here... Tested a few dynamic object field lookups manually. We do not need this logic here, dynamicFieldConnection returns the correct data for us to pass back into dynamicField(name)

Reply:
Was looking into the wrong thing. Please ignore - will investigate more closely

Reply:
Ok, diligently looked into this. Yes to resolve DOF we need to make some more lookups. I didn't think it was an issue that we do this for a dynamic object field since the indexerv2 + jsonrpc on fullnode implement dynamic field lookups the same way. 

For a dynamic object field with DynamicFieldName of type tag `0x2::kiosk::Item` and the bcs of the name's value (an undecorated MoveValue) such as `Some(Struct(Runtime([Struct(Runtime([Address(34b02bc7550982e3946265e0346f0fbfc29b2972e35ad0a2ea2d187aad5585fc)]))])))`, if I attempt to derive dynamic id, I get `0xaa40a57b36e714c1e0a5693552bfb9f6e9bfb9a439f82e67d50b6b08cf433876` which does not correspond to either the object_id of the wrapper struct or the object wrapped as child.

When I modify the TypeTag to `0x2::dynamic_object_field::Wrapper<0x2::kiosk::Item>`, without changing the bcs of the name's value in the graphql implementation, the derived dynamic id points to the wrapper struct's id. Fetching this object from db finally yields the wrapped object's id, which can be used to query for the object. This is consistent w/ the jsonrpc implementation.

Not that we have to follow precedent, but it doesn't seem possible to derive the dynamic object field id from the typing and serialized data of the inner T of F<Wrapper<T>, V>. But I'm not really a fan of separate APIs for DynamicField and DynamicObjectField since we basically push this work onto the client. Especially since dynamicFieldConnection resolves F<Wrapper<T>, V> as below to keep it in line w/ how we represent dynamic fields
```
DynamicField {
  name -> type and bcs of T
  value -> the child object
```

Reply:
> I didn't think it was an issue that we do this for a dynamic object field since the indexerv2 + jsonrpc on fullnode implement dynamic field lookups the same way.

I find this a kind of strange argument, because if we implemented everything the same way as they are implemented for JSON-RPC and fullnode, then ...why are we re-implementing anything?  I think you get the strangeness because you mention that we don't have to follow precedent, but I think this is worth calling out!

The existing JSON-RPC implementation is broken in a way that we wanted to fix, here's a snippet from an earlier conversation we had on Slack about it:

> While discussing the Display DSL for dynamic fields we noticed a small issue with the way the existing dynamic field RPCs work:
>
> In Move, it's possible to have a dynamic field and a dynamic object field that share the same key but map to different values, because they are accessed via different functions, but in the RPC they are collapsed into one API (getDynamicFieldObject).
>
> This is implemented using a secondary index for dynamic fields which stores both kinds of dynamic field in the same table, so the keys will overlap and that index will only record one of those mappings (different from the state on chain it's representing).
>
> It's basically impossible to make the current RPC behaviour match the on-chain behaviour without breaking backwards compatibility, and given this is quite a niche case, I don't think it's worth doing that, but we should fix this behaviour in the redesign.

The data representation issue was fixed in the Indexer V2 schema, but its use in JSON-RPC needs to preserve this behaviour, so it doesn't expose the fix in the interface, but **we should for GraphQL**.

> it doesn't seem possible to derive the dynamic object field id

Yes, this is correct: it's not possible from knowing the parent ID, name type and name BCS whether that corresponds to a dynamic field or a dynamic object field.  But if you think about it, thinking that this is possible is the source of the bug in the first place: For a single parent, type and value there could be both a dynamic field value and a dynamic object field value.

> But I'm not really a fan of separate APIs for DynamicField and DynamicObjectField since we basically push this work onto the client.

We are pushing the work onto the client, because if the client is making the request for a specific field, by name, it should know whether it is a dynamic field or a dynamic object field.  As the implementor of the API, we don't know that, and so we are making everybody's performance worse to compensate for that, which is not a good trade off.

> Especially since dynamicFieldConnection resolves F<Wrapper, V> as below to keep it in line w/ how we represent dynamic fields

I don't think `dynamicField` and a hypothetical `dynamicObjectField` should have a different representation than `dynamicFieldConnection` for its output -- they should all use `DynamicField`, whose `value` field also indicates whether the result is a dynamic field or a dynamic object field.

In fact, one of the reasons I find the current approach concerning is that `dynamicFieldConnection` could return two `DynamicField`s with the same name (a dynamic field and a dynamic object field) while the point lookup API will not be able to look up the dynamic object field.

Reply:
> For a single parent, type and value there could be both a dynamic field value and a dynamic object field value.

Got it, this is the critical piece that I did not consider ... thanks for connecting the dots! I'm 200% onboard